### DISCUSSION TITLE
Is there some level of memoization?

### DISCUSSION THREAD

**Discussion by JoeDuncko** [css-hooks/css-hooks#38] [General]
Hi! I really love the idea of this project. I personally really like using inline styles where possible, and this seems to be a great way to extend that functionality for some very common use cases.

One of the biggest arguments against inline styles in react is that because they are typically written as an object inline, it can break memoization of components deeper in the stack because the styles produce a new object every render.

Does this library have an answer to that argument? Or is it purely focused on adding these additional styles hooks?

Thanks!

**Comment by nsaunders** [comment]
Hi @JoeDuncko, welcome, and thanks for the kind words and great question! I think a lot of people are uncertain about this issue. Including me at times! So I asked about it [here](https://twitter.com/agilecoder/status/1710397947710779732), and @ziir jumped in to set the record straight:

> Memoization is unnecessary if you're just passing values to host (or "native") components. Only useful if computationally expensive, or stable references matter for some reason.

I later found a number of resources that agree with this answer. One I find particularly useful is [Death by a thousand useCallbacks](https://royi-codes.vercel.app/thousand-usecallbacks/), which offers several guidelines to help determine when to memoize a value.

It's worth noting that CSS Hooks' style object flattening algorithm does involve a few operations like traversing nested style objects, reading/assigning property values, and string interpolation; but these are all pure computations (no DOM interaction or other potentially costly effects); so I doubt it belongs in that "computationally expensive" category (although benchmarks are welcome).

I hope this helps, and please let me know if I've missed anything.

**Reply by JoeDuncko** [reply]
Thanks @nsaunders! I think that the Twitter replies make sense for that exact scenario, where 

`<div style={useMemo(() => ({ color: "blue" }), [])}>`

is just a DOM element, which would only rerender if something else triggered a rerender and is presumably cheap to render.

EDIT: Just to clarify, I was not particularly concerned with CSS Hooks' perf, and trying to babysit it with memoization. I agree that that's probably overkill.

My question is more about the following scenario:

Say you have `ExpensiveComponent` that can take a `style` property. Maybe it renders a fancy map or something. `ExpensiveComponent` is wrapped in `React.memo()`, so that it only rerenders if its props actually change.

Maybe it looks like this:

```js
function App() {
    const [someState, setSomeState] = useState(0);

    return <div>
        <ExpensiveComponent style={css({style: "value"})} />
        <CheapComponent someState={someState} />
    </div>
}
```

When `someState` changes, my understanding is that `ExpensiveComponent` will rerender even if its memoized because `css({style: "value"})` presumably gives us a different object output each time...

Though as I type that, I think the actual issue here is that `{style: "value"}` is an object that gets re-created every render, and if `css({style: "value"})` actually outputs a STRING, and it produces the same string every time given the same shaped object, `ExpensiveComponent`'s memoization would trigger and everything would work as desired.

Does css-hooks work the way I'm hoping it does then? And does my thinking line up correctly? If so, I'm very hype to use this.

Thanks!

----

It's worth noting that there are many ways of dealing with the above scenario. We could store our state in a third party location, or we could relocate the state. But ignoring that (as we could continue building out this example to make this base situation more valid), we could also use other css-in-js solutions or tailwind, which deal with this issue for us.

**Comment by nsaunders** [comment]
Hi @JoeDuncko, thanks for helping me to understand the scenario. Your assessment is totally correct.

Out of curiosity, have you considered the second `arePropsEqual` parameter of the `memo` function? e.g.

```js
const ExpensiveComponent = memo(
  ExpensiveComponentImpl,
  (a, b) => JSON.stringify(a.style) === JSON.stringify(b.style) // todo compare other props as needed
);
```

If that doesn't work or is too inconvenient, your other idea would work, too. Because you own your own `css` function, you can just create a "wrapper" function that returns a string instead of an object. Something like this:

```typescript
const [hooks, cssImpl] = createHooks({
  // hooks config
});

export const css = (properties: Parameters<typeof cssImpl>[0]) => {
  const styleObject = cssImpl(properties);
  const styleString = styleObjectToString(styleObject); // see https://www.npmjs.com/search?q=style%20object%20to%20string
  return styleString;
};
```

The main disadvantage I see in the latter approach is that merging styles is less straightforward, but honestly that may be a small price to pay (especially depending on your use case).

I guess one other variation on this idea would be to stringify the styles on a one-off basis, e.g.

```tsx
function App() {
    const [someState, setSomeState] = useState(0);

    return <div>
        <ExpensiveComponent style={styleObjectToString(css({style: "value"}))} />
        <CheapComponent someState={someState} />
    </div>
}
```

Please let me know if any of these suggestions make sense for you. And I should note that I'm open to suggestions as well, in case any others come to mind. :-)
